tokio/src/sync/set_once.rs

```diff
@@ -0,0 +1,409 @@
+    /// Returns `None` if the cell is empty.
+    pub fn into_inner(self) -> Option<T> {
+        // TODO: Use get_mut()
+        let value_set = self.value_set.load(Ordering::Relaxed);
+
+        if value_set {
+            // Since we have taken ownership of self, its drop implementation
+            // will be called by the end of this function, to prevent a double
+            // free we will set the value_set to false so that the drop
+            // implementation does not try to drop the value again.
+            self.value_set.store(false, Ordering::Relaxed);
+
+            // SAFETY: The SetOnce is currently initialized, we can assume the
+            // value is initialized and return that, when we return the value
+            // we give the drop handler to the return scope.
+            Some(unsafe { self.value.with_mut(|ptr| ptr::read(ptr).assume_init()) })
+        } else {
+            None
+        }
+    }
+
+    /// Waits until set is called. The future returned will keep blocking until
+    /// the `SetOnce` is initialized.
+    ///
+    /// If the `SetOnce` is already initialized, it will return the value
+    /// immediately.
+    ///
+    /// # Note
+    ///
+    /// This will keep waiting until the `SetOnce` is initialized, so it
+    /// should be used with care to avoid blocking the current task
+    /// indefinitely.
+    pub async fn wait(&self) -> &T {
+        loop {
+            let notify_fut = self.notify.notified();
+            pin!(notify_fut);
+            // take the lock because we're reading value_set
+            let guard = self.lock.lock();
+
+            if self.value_set.load(Ordering::Relaxed) {
+                // SAFETY: the state is initialized
+                return unsafe { self.get_unchecked() };
+            }
+
+            drop(guard);
+
+            // wait until the value is set
+            notify_fut.await;
+        }
+    }
+}
+
+// Since `get` gives us access to immutable references of the SetOnce, SetOnce
+// can only be Sync if T is Sync, otherwise SetOnce would allow sharing
+// references of !Sync values across threads. We need T to be Send in order for
+// SetOnce to by Sync because we can use `set` on `&SetOnce<T>` to send values
+// (of type T) across threads.
+unsafe impl<T: Sync + Send> Sync for SetOnce<T> {}
+
+// Access to SetOnce's value is guarded by the Atomic boolean flag
+// and atomic operations on `value_set`, so as long as T itself is Send
+// it's safe to send it to another thread
+unsafe impl<T: Send> Send for SetOnce<T> {}
+
+/// Error that can be returned from [`SetOnce::set`].
+///
+/// This error means that the `SetOnce` was already initialized when
+/// set was called
+///
+/// [`SetOnce::set`]: crate::sync::SetOnce::set
+#[derive(Debug, PartialEq, Eq)]
+pub struct SetOnceError<T> {
+    /// The cell was already initialized when [`SetOnce::set`] was called.
+    ///
+    /// [`SetOnce::set`]: crate::sync::OnceCell::set
+    val: T,
+}
```

Comment:
I think it's fine to just do this:
```suggestion
pub struct SetOnceError<T>(pub T);
```
That lets you get rid of the `From` impl and write `SetOnceError(value)` instead of `SetOnceError::from(value)`. It also allows the end-use to get the value out of the error, which is the primary reason we return the value in the error.

Reply:
Removed From impl and added the tuple strucz